tools/lambda-promtail/lambda-promtail/promtail.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	timeout    = 5 * time.Second
+	minBackoff = 100 * time.Millisecond
+	maxBackoff = 30 * time.Second
+	maxRetries = 10
+
+	reservedLabelTenantID = "__tenant_id__"
+
+	userAgent = "lambda-promtail"
+)
+
+type entry struct {
+	labels model.LabelSet
+	entry  logproto.Entry
+}
+
+type batch struct {
+	streams map[string]*logproto.Stream
+	size    int
+}
+
+func newBatch(ctx context.Context, entries ...entry) (*batch, error) {
+	b := &batch{
+		streams: map[string]*logproto.Stream{},
+	}
+
+	for _, entry := range entries {
+		err := b.add(ctx, entry)
+		return b, err
+	}
+
+	return b, nil
+}
+
+func (b *batch) add(ctx context.Context, e entry) error {
+	labels := labelsMapToString(e.labels, reservedLabelTenantID)
+	if stream, ok := b.streams[labels]; ok {
+		stream.Entries = append(stream.Entries, e.entry)
+
+		b.size += stream.Size()
```

Comment:
This `Size()` method doesn't actually measure what we're expecting (it's part of the generated protobuf). Instead, I'd keep track of the line's bytel length as they're added, like so:
https://github.com/grafana/loki/blob/main/clients/pkg/promtail/client/batch.go#L44-L45

Reply:
I see, will fix that